### PR TITLE
Avoid race between image pull and garbage collect

### PR DESCRIPTION
**Pull request by anguslees**
Hold the pullLock while pulling regular container images, not just the
network container.

Also: previously, the locks were grabbed at too-low a layer in the
callstack, leaving a (narrow) window where images could be pulled and
then garbage collected before being used.

This change takes the existing pullLock a layer higher in the call stack
so the critical regions encompass the entire find+delete and pull+use
operation in both cases.

### CODE REVIEW

**Comment by brendandburns** [comment]
The problem is really underlying in Docker.  There is really no good way to know if an image is unreachable.

I'm ok with only GC-ing images where the Created is > some window.

This whole thing is a giant hack, but it's hard to see how to resolve it without better support inside of Docker.

File: pkg/kubelet/kubelet.go [hunk 1]
```diff
@@ -814,6 +818,8 @@ func (kl *Kubelet) syncPod(pod *api.BoundPod, dockerContainers dockertools.Docke
 		if err != nil {
 			glog.Errorf("Couldn't make a ref to pod %v, container %v: '%v'", pod.Name, container.Name, err)
 		}
+		kl.pullLock.RLock()
+		defer kl.pullLock.RUnlock()
```
Comment (brendandburns) [inline pkg/kubelet/kubelet.go]: this won't work right.  the defer is scoped to the function.  The right answer is to make this use the pullImage function.

**Comment by brendandburns** [comment]
with regard to lifting the scope of the lock, I think it might actually be necessary, you could imagine:
```
thread-1:                                                       thread-2:
lock
list := list all unused images, "foo" is unused
unlock
                                                                pull "bar" that depends on "foo", "foo" is no longer dangling
delete "foo"
```

This might explain some of the pull flakiness that we're seeing.

**Comment by brendandburns** [comment]
I think this is an important bug to get fixed, so I'm going to manually merge this with a move of the ReadUnlock.  We can re-adjust it later.

**Comment by brendandburns** [comment]
I just reverted this.  e2e completely blew up after it went in.

e2e has been a little shaky in general, but this seemed worse, I'd like to take this a little bit more slowly so that we can understand exactly what is going on.

There is clearly a problem here (esp. with the container image pulls) I'm taking it as a personal P0 to get it sorted out.

**Comment by anguslees** [comment]
I just sent a new iteration of my patch - I hadn't seen your comment just now and I haven't run the e2e tests.

The race you mention above is exactly the one I had in mind when writing this change, and I don't think your amended version of my patch is correct.  In particular, the wlock is not held between "getUnusedImages" and "actually delete the images", and the rlock is not held between "pull image" and "runContainer" (ie: remove it from the docker "dangling" list used by getUnusedImages).

**Comment by brendandburns** [comment]
Just sent https://github.com/GoogleCloudPlatform/kubernetes/pull/3264

which only locks around the docker pull.

As I understand it, one the pull is complete, the image is "tagged" and so therefore it shouldn't be returned by the "dangling=true" filter, so I _think_ we only need to guard the pull against the list images query.

I will admit that I really don't fully understand how docker determines "dangling" so there could still be problems.  #3264 is certainly better than the state of HEAD, and we can incrementally move to add protection for the other race that you are concerned with (and make sure e2e continues to be clean...)

**Pull request merged**